File: ts/packages/luzid-sdk/src/api/validator.ts
import {
  ValidatorInfo,
  ValidatorOperation,
  ValidatorOpsRequest,
  ValidatorOpsResponse,
  ValidatorStats,
} from '@luzid/grpc'

import type { LuzidGrpcClient } from '@luzid/grpc-client'
import { Successful, unwrap } from '../core/utils'
import { ValidatorStatus } from '@luzid/grpc'

export class LuzidValidator {
  constructor(private readonly client: LuzidGrpcClient) {}

  /**
   * Performs an operation on the validator.
   *
   * @param op - The operation to perform (start|stop|restart)
   *
   * @private
   */
  private async validatorOps(
    op: ValidatorOperation
  ): Promise<Successful<ValidatorOpsResponse>> {
    const req: ValidatorOpsRequest = { op }
    const res = await this.client.validator.validatorOps(req)
    return unwrap(res, 'Luzid validator.validatorOps')
  }

  /**
   * Starts the validator.
   */
  start(): Promise<Successful<ValidatorOpsResponse>> {
    return this.validatorOps(ValidatorOperation.Start)
  }

  /**
   * Stops the validator.
   */
  stop(): Promise<Successful<ValidatorOpsResponse>> {
    return this.validatorOps(ValidatorOperation.Stop)
  }

  /**
   * Restarts the validator.
   */
  restart(): Promise<Successful<ValidatorOpsResponse>> {
    return this.validatorOps(ValidatorOperation.Restart)
  }

  /*
   * Subscribes to validator info updates.
   * This usually is emitted once per validator start.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  subValidatorInfo({ emitCurrent = false } = {}): AsyncIterable<ValidatorInfo> {
    return this.client.validator.subValidatorInfo({ emitCurrent })
  }

  /**
   * Retrieves the current validator info or the earliest available.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  async getValidatorInfo({ emitCurrent = true } = {}): Promise<ValidatorInfo> {
    const iter = this.client.validator.subValidatorInfo({ emitCurrent })
    for await (const info of iter) {
      return info
    }
    throw new Error('Missing ValidatorInfo')
  }

  /**
   * Subscribes to validator status updates.
   * This usually is emitted once per validator start or stop.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  subValidatorStatus({
    emitCurrent = false,
  } = {}): AsyncIterable<ValidatorStatus> {
    return this.client.validator.subValidatorStatus({ emitCurrent })
  }

  /**
   * Retrieves the current validator status or the earliest available.
   * This is emitted once per validator start or stop.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  async getValidatorStatus({
    emitCurrent = true,
  } = {}): Promise<ValidatorStatus> {
    const iter = this.client.validator.subValidatorStatus({ emitCurrent })
    for await (const status of iter) {
      return status
    }
    throw new Error('Missing ValidatorStatus')
  }

  /**
   * Subscribes to validator stats updates.
   * This is emitted per slot.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  subValidatorStats({
    emitCurrent = false,
  } = {}): AsyncIterable<ValidatorStats> {
    return this.client.validator.subValidatorStats({ emitCurrent })
  }

  /**
   * Retrieves the current validator stats or the earliest available.
   * This is emitted per slot.
   *
   * @param opts
   * @param opts.emitCurrent - Whether to emit the current value.
   */
  async getValidatorStats({
    emitCurrent = true,
  } = {}): Promise<ValidatorStats> {
    const iter = this.client.validator.subValidatorStats({ emitCurrent })
    for await (const stats of iter) {
      return stats
    }
    throw new Error('Missing ValidatorStats')
  }
}

File: dart/luzid-sdk/lib/src/api/validator.dart
import 'package:luzid_grpc/luzid_grpc.dart'
    hide ValidatorStatus, ValidatorInfo, ValidatorStats;
import 'package:luzid_grpc_client/luzid_grpc_client.dart';
import 'package:luzid_sdk/src/api-types/validator_info.dart';
import 'package:luzid_sdk/src/api-types/validator_stats.dart';
import 'package:luzid_sdk/src/api-types/validator_status.dart';
import 'package:luzid_sdk/src/core/utils.dart';

export 'package:luzid_sdk/src/api-types/validator_info.dart';
export 'package:luzid_sdk/src/api-types/validator_stats.dart';
export 'package:luzid_sdk/src/api-types/validator_status.dart';

class LuzidValidator {
  final ValidatorClient _client;

  LuzidValidator(LuzidGrpcClient client) : _client = client.validator;

  /// Performs an operation on the validator.
  Future<ValidatorOpsResponse> _validatorOps(ValidatorOperation op) async {
    final res = await _client.validatorOps(ValidatorOpsRequest(op: op));
    return unwrap(res, 'Luzid validator.validatorOps');
  }

  /// Starts the validator.
  Future<ValidatorOpsResponse> start() {
    return _validatorOps(ValidatorOperation.Start);
  }

  /// Stops the validator.
  Future<ValidatorOpsResponse> stop() {
    return _validatorOps(ValidatorOperation.Stop);
  }

  /// Restarts the validator.
  Future<ValidatorOpsResponse> restart() {
    return _validatorOps(ValidatorOperation.Restart);
  }

  Stream<ValidatorStatus> subValidatorStatus({bool emitCurrent = false}) {
    return _client.subValidatorStatus(emitCurrent).map((rpc) {
      return ValidatorStatus.from(rpc);
    });
  }

  Stream<ValidatorInfo> subValidatorInfo({bool emitCurrent = false}) {
    return _client.subValidatorInfo(emitCurrent).map((rpc) {
      return ValidatorInfo.from(rpc);
    });
  }

  Stream<ValidatorStats> subValidatorStats() {
    return _client.subValidatorStats().map((rpc) {
      return ValidatorStats.from(rpc);
    });
  }
}
